### PR TITLE
Only add Content-Type if not specified in headers. Fixes #20046

### LINKED ISSUE #20046: URI module - pass JSON body without setting Content-Type header to 'application/json'
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
uri module

##### ANSIBLE VERSION
```
$ ansible --version
ansible 2.3.0 (devel c1f5432cc0) last updated 2017/01/09 14:36:42 (GMT +100)
  config file = /Users/nbailey/automation/roles-linux/ambari_server/tests/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
`hash_behaviour=merge` is the only even potentially relevant one.

##### OS / ENVIRONMENT
n/a

##### SUMMARY

I've hit an issue a couple of times where an API wants JSON formatted data but a Content-Type header set to 'application/x-www-form-urlencoded'.

I can't find a convenient way to pass a JSON body to the URI module, but avoid setting the header to `application/json`.

If I set `body_format` to `raw`, I can set the body to a string that contains valid JSON... But if I use a lookup against a Jinja template, the module gets handed a YAML hash instead, and I get an error ("An unknown error occurred: sendall() argument 1 must be string or buffer, not list").

##### STEPS TO REPRODUCE

I'm trying to interact with the Ambari Server API (https://github.com/apache/ambari/blob/trunk/ambari-server/docs/api/v1/index.md), doing the equivalent of this curl:

```
$ curl --user admin:admin -H 'X-Requested-By: ambari' -X POST -d '[{"Event": {"specs": [{"names": "<redacted>,<redacted>", "principal_type": "groups", "sync_type": "specific"}]}}]' http://192.168.2.20:8080/api/v1/ldap_sync_events
{
  "resources" : [
    {
      "href" : "http://192.168.2.20:8080/api/v1/ldap_sync_events/6",
      "Event" : {
        "id" : 6
      }
    }
  ]
```
Task using `body_format: json`:

```
- name: Synchronise LDAP groups and users.
  uri:
    url: "{{ ambari_server.endpoint | default(inventory_hostname) }}/api/v1/ldap_sync_events"
    headers:
      X-Requested-By: "ambari"
      Content-Type: "application/x-www-form-urlencoded"
    method: POST
    user: "{{ ambari_server.admin_user | default('admin') }}"
    password: "{{ ambari_server.admin_password | default('admin') }}"
    force_basic_auth: True
    body_format: raw
    body: "{{ lookup('template', 'ldap_sync.json.j2') }}"
    status_code: 201
    return_content: True
  when: ambari_server is defined and ambari_server.ldap is defined
```

This passes the JSON blob happily, but the server doesn't like it. 

Response: "Status code was not [201]: HTTP Error 500: java.lang.IllegalStateException: Expected a string but was BEGIN_ARRAY at line 1 column 2

Running the task with `body_format: raw` to avoid getting `application/json` set gives me an Ansible-side error: "Status code was not [201]: An unknown error occurred: sendall() argument 1 must be string or buffer, not list". 

I don't understand how Ansible formats parameters before handing them off to modules very well. I think what's happening here is that the module errors on getting a YAML hash rather than a string. The `body_format: json` parameter expects JSON and uses `json.dumps()` on that parameter before passing it on to whatever makes the HTTP request, but the module code that handles things with a `raw` format does not. 

What does work? I don't know if this is a good or widely useful solution, but I just set the uri module to only set a Content-Type header if there wasn't already one:

```
    # grabbed from https://github.com/ansible/ansible-modules-core/blob/devel/network/basics/uri.py#L398
    # commit d749ec91e513c40a50b539a7eaccf29c31dda4d4
    if body_format == 'json':
        # Encode the body unless its a string, then assume it is pre-formatted JSON
        if not isinstance(body, basestring):
            body = json.dumps(body)
        if 'Content-Type' not in dict_headers:
            dict_headers['Content-Type'] = 'application/json'
```

Is this is a useful default behaviour for the module? If so, I'm happy to put a PR in for it.

If not, is there another way I can achieve the same effect?

### PR DESCRIPTION
**Pull request by sivel**
##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
uri

##### ANSIBLE VERSION

```
v2.3
```

##### SUMMARY

Previously, if `body_format: json` was specified, `Content-Type: application/json` was forcefully set.  If a user has set `Content-Type` via `headers` already, do not override, as they may need a custom `Content-Type`

Fixes #20046

### CODE REVIEW

**Comment by sivel** [comment]
FYI, I am unable to determine why the `ubuntu1604py3` test is failing.  It looks like it is actually failing on a prior task, where we do a `wait_for` on the `testserver.py`.  Running the `ubuntu1604py3` tests locally are succeeding, and restarting the build didn't seem to fix it.

**Comment by bcoca** [comment]
I would add note in documentation for the headers option: since 2.3 it is possible for this to override the ....

File: test/integration/targets/uri/tasks/main.yml [hunk 1]
```diff
@@ -280,3 +280,16 @@
     status_code: 202
     method: POST
     body: foo
+
+- name: Validate body_format json does not override content-type
```
Comment (bcoca) [inline test/integration/targets/uri/tasks/main.yml]: add (works in versions >=2.3)

Comment (sivel) [reply test/integration/targets/uri/tasks/main.yml]: Will do

**Comment by sivel** [comment]
Documentation updated for this change

**Pull request merged**